moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java

```diff
@@ -297,9 +297,9 @@ public void bind(Moshi moshi, JsonAdapter.Factory factory) {
           Type type = ((ParameterizedType) parameterTypes[i]).getActualTypeArguments()[0];
           Set<? extends Annotation> jsonAnnotations = jsonAnnotations(parameterAnnotations[i]);
           jsonAdapters[i - adaptersOffset] =
-              Types.equals(this.type, type) && annotations.equals(jsonAnnotations)
-                  ? moshi.nextAdapter(factory, type, jsonAnnotations)
-                  : moshi.adapter(type, jsonAnnotations);
+              Types.equals(Types.canonicalize(this.type), Types.canonicalize(type))
```

Comment:
Yes, the test fails without it.
Cool, I'll send a PR to replace this one, then.